services/src/main/java/io/grpc/services/ChannelzService.java

```
@@ -0,0 +1,167 @@
+import io.grpc.stub.StreamObserver;
+
+public class ChannelzService extends ChannelzGrpc.ChannelzImplBase {
+  private static final int DEFAULT_MAX_PAGE_SIZE = 25;
+  private final Channelz channelz;
+  private final int maxPageSize;
+
+  public ChannelzService() {
+    this(Channelz.instance(), DEFAULT_MAX_PAGE_SIZE);
+  }
+
+  @VisibleForTesting
+  ChannelzService(Channelz channelz, int maxPageSize) {
+    this.channelz = channelz;
+    this.maxPageSize = maxPageSize;
+  }
+
+  /** Returns top level channel aka {@link io.grpc.internal.ManagedChannelImpl}. */
+  @Override
+  public void getTopChannels(
+      GetTopChannelsRequest request, StreamObserver<GetTopChannelsResponse> responseObserver) {
+    try {
+      Channelz.RootChannelList rootChannels
+          = channelz.getRootChannels(request.getStartChannelId(), maxPageSize);
+      GetTopChannelsResponse.Builder responseBuilder = GetTopChannelsResponse
+          .newBuilder()
+          .setEnd(rootChannels.end);
+      for (Instrumented<ChannelStats> c : rootChannels.channels) {
+        responseBuilder.addChannel(ChannelzProtoUtil.toChannel(c));
+      }
+      responseObserver.onNext(responseBuilder.build());
+      responseObserver.onCompleted();
+    } catch (Throwable t) {
+      responseObserver.onError(t);
+    }
+  }
+
+  /** Returns a top level channel aka {@link io.grpc.internal.ManagedChannelImpl}. */
+  @Override
+  public void getChannel(
+      GetChannelRequest request, StreamObserver<GetChannelResponse> responseObserver) {
+    try {
+      Instrumented<ChannelStats> s = channelz.getRootChannel(request.getChannelId());
+      if (s == null) {
+        responseObserver.onError(Status.NOT_FOUND.asRuntimeException());
+        return;
+      }
+
+      responseObserver.onNext(
+          GetChannelResponse
+              .newBuilder()
+              .setChannel(ChannelzProtoUtil.toChannel(s))
+              .build());
+      responseObserver.onCompleted();
+    } catch (Throwable t) {
```

same here.  It would be better to fail hard, rather than silently.

---

done